src/mpfao/tdympfao.c

```diff
@@ -1950,6 +1950,96 @@ static PetscErrorCode ComputeFacePermeabilityTensor(TDyMPFAO *mpfao,
  PetscFunctionReturn(0);
 }
 
+/// For the TPF approach, compute the value of permeability at face center that
+/// is used in computing the discretization of gravity term.
+///
+/// - First, compute permeability scalar values from the up and down cell permeability
+///   tensor using PFLOTRAN's approach
+/// - Second, compute distance-weighted harmonic-average of scalar permeabilities
+///
+/// @param [in] tdy A TDy struct
+/// @param [in] dim dimension of the problem
+/// @param [in] face_id ID of the face
+/// @param [out] *Kface_value Permeability value at the face
+/// @returns 0 on success, or a non-zero error code on failure
+PetscErrorCode ComputeFacePeremabilityValueTPF(TDyMPFAO *mpfao, MaterialProp *matprop, PetscInt dim, PetscInt face_id, PetscReal *Kface_value) {
+
+  PetscFunctionBegin;
+
+  PetscErrorCode ierr;
+
+  PetscReal u_up2dn[dim];
+  ierr = ComputeUpDownUnitVector(mpfao, face_id, u_up2dn); CHKERRQ(ierr);
+
+  PetscReal Kup[dim*dim], Kdn[dim*dim];
+  ierr = ExtractUpAndDownPermeabilityTensors(mpfao, matprop, face_id, dim, Kup, Kdn); CHKERRQ(ierr);
+
+  PetscReal Kup_value = 0.0, Kdn_value = 0.0;
+  for (PetscInt kk=0; kk<dim; kk++) {
+    Kup_value += pow(u_up2dn[kk],2.0)/Kup[kk*dim + kk];
+    Kdn_value += pow(u_up2dn[kk],2.0)/Kdn[kk*dim + kk];
+  }
+
+  Kup_value = 1.0/Kup_value;
+  Kdn_value = 1.0/Kdn_value;
+
+  PetscReal dist_up, dist_dn;
+  ierr = ComputeUpAndDownDist(mpfao, face_id, &dist_up, &dist_dn); CHKERRQ(ierr);
+
+  PetscReal wt_up = dist_up / (dist_up + dist_dn);
+
+  *Kface_value = (Kup_value*Kdn_value)/(wt_up*Kdn_value + (1.0-wt_up)*Kup_value);
+
+  PetscFunctionReturn(0);
+}
+
+/// For the MPFAO approach, compute the value of permeability at face center that
+/// is used in computing the discretization of gravity term.
+///
+/// - First, permeability tensor at the face is computed by harmonically averaging
+///    permeability tensors of cells share the face
+/// - Second, the face permeability tensor is projected along the unit normal
+///   along the line joining up and down cells
+/// - Third, dot producot of the projected face permeability vector and unit
+///   normal to the face is computed
+///
+/// @param [in] tdy A TDy struct
+/// @param [in] dim dimension of the problem
+/// @param [in] face_id ID of the face
+/// @param [out] *k_face_value Permeability value at the face
+/// @returns 0 on success, or a non-zero error code on failure
+PetscErrorCode ComputeFacePeremabilityValueMPFAO(TDyMPFAO *mpfao, MaterialProp *matprop, PetscInt dim, PetscInt face_id, PetscReal *Kface_value) {
```

Comment:
Peremability -> Permeability

Reply:
@bishtgautam , can you fix this typo? It's not a big deal, but it's easier to fix now than later.